toolchain/lower/file_context.cpp

```diff
@@ -190,49 +190,31 @@ auto FileContext::GetOrCreateFunction(SemIR::FunctionId function_id,
   return result;
 }
 
-auto FileContext::BuildFunctionDecl(SemIR::FunctionId function_id,
-                                    SemIR::SpecificId specific_id)
-    -> llvm::Function* {
-  const auto& function = sem_ir().functions().Get(function_id);
-
-  // Don't lower generic functions. Note that associated functions in interfaces
-  // have `Self` in scope, so are implicitly generic functions.
-  if (function.generic_id.has_value() && !specific_id.has_value()) {
-    return nullptr;
-  }
-
-  // Don't lower builtins.
-  if (function.builtin_function_kind != SemIR::BuiltinFunctionKind::None) {
-    return nullptr;
+auto FileContext::BuildFunctionTypeInfo(
+    const SemIR::Function& function, SemIR::SpecificId specific_id,
+    llvm::ArrayRef<SemIR::InstId> implicit_param_patterns,
+    llvm::ArrayRef<SemIR::InstId> param_patterns,
+    const SemIR::ReturnTypeInfo& return_info,
+    llvm::SmallVector<llvm::Type*>& param_types,
+    llvm::SmallVector<SemIR::InstId>& param_inst_ids, llvm::Type*& return_type,
+    SemIR::InstId& return_param_id) -> void {
+  return_type = nullptr;
```

Comment:
Maybe we should also clear the other out params? Or document and assume they are all initialized correctly? Just clearing one seems inconsistent either way, and not like something we'd want to accidentally end up depending on.

Reply:
overlaps with https://github.com/carbon-language/carbon-lang/pull/5066#discussion_r1980035814, discussed there